### ISSUE TITLE
Is there a layer "up_sampling2d_5"?

### ISSUE THREAD

**Issue by xuyingxiao** [chungtd89/grad-cam-heat-map-membrane-segmentation-u-net#1]
Hi,
you mentioned last_conv_layer = model.get_layer('up_sampling2d_5') with shape  (?, 32, 32, 1024), is this layer should be up_sampling2d_1? I didn't find up_sampling2d_5 in unet model.
And why you choose this layer but not the last conv layer in decoder "up_sampling2d_4" or the last conv layer in encoder "conv2d_10"? 
Thanks for any reply!

**Comment by chungtd89** [comment]
Nope, at step: Generate Heatmap, if print out model summary again, you will see up_sampling2d_5 - up_sampling2d_8 corresponding to up_sampling2d_1-up_sampling2d_4. 
model1 for training: up_sampling2d_1-up_sampling2d_4
model2 for prediction: up_sampling2d_5-up_sampling2d_8

Choosing different layers will affect the grads_shape and thus, conv_layer_output_value / heatmap shape.

**Comment by xuyingxiao** [comment]
Dear chungtd89:


Thanks for your reply, I got the heatmap with different layers of my model. But there is a problem I met when I wanna to try it on the road extraction task with model Dlinknet(add dilation block to unet,  https://github.com/zlkanata/DeepGlobe-Road-Extraction-Challenge), why the heatmap from the last conv of encoder"encoder4" and the first conv of decoder block "decoder1"  so different？ 


heatmap from encoder4




heatmap with decoder1


For segmentation task,  I'm NOT sure the backward scalar. Is that right?


Maybe there is something mistake because of my incomplete knowledge. Can you have a look and give some suggestions? Appreciate for any reply.


If you have a little time,please see details in the attachment(my code and pretrained weights with pytorch), Thanks again.






Yingxiao xu


| |
徐樱笑
|
|
yingxiao_xu@163.com
|
签名由网易邮箱大师定制
On 5/7/2020 21:59，chungtd89<notifications@github.com> wrote：

Nope, at step: Generate Heatmap, if print out model summary again, you will see up_sampling2d_5 - up_sampling2d_8 corresponding to up_sampling2d_1-up_sampling2d_4.
model1 for training: up_sampling2d_1-up_sampling2d_4
model2 for prediction: up_sampling2d_5-up_sampling2d_8

Choosing different layers will affect the grads_shape and thus, conv_layer_output_value / heatmap shape.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub, or unsubscribe.

从网易邮箱大师发来的云附件
code.zip(113.1MB,2020年5月22日 23:5 到期)
下载

**Comment by xuyingxiao** [comment]
why different layers have so different heatmap?

**Comment by chungtd89** [comment]
Because heatmap size is different on different layers: see last_conv_layer.output[0]_shape, this affects heatmap calculation.

**Comment by xuyingxiao** [comment]
I noticed that the heatmap is applied for prediction result, why prediction result rather than original image? Grad cam is proposed to visualize the activation map of the network, understand and analyze the learning situation of the network,. 
How do I know what I pay attention to in the network model?

**Comment by chungtd89** [comment]
Original images are used for training. If you apply heatmap to original image, it is ok for visualization purpose. But if you use heatmap info for training, it may affect prediction results.
Heatmap applying on prediction results help you to understand better the outcome of the prediction. E.g., if prediction result is just a gray-scale image, you may not notice important features in the image that make the image to be classified as a particular object.
What should you pay attention to the network model:
- For a repository, make sure you can run it first. If cannot run, it takes a lot of time to further investigate.
- For a network, it depends on your needs. If you want to visualize features of generated image, go to check heatmap portion. If you want to make a new model, go and study model and codes related to it.
